src/mjolnir/graphtilebuilder.cc

```diff
@@ -1258,5 +1265,63 @@ void GraphTileBuilder::UpdatePredictedSpeeds(const std::vector<DirectedEdge>& di
   }
 }
 
+void GraphTileBuilder::AddLandmark(const GraphId& edge_id, const Landmark& landmark) {
+  // first thing is check that the edge id makes sense, tile id should match and edges_.size() >
+  // edge_id.id if not throw a runtime error. otherwise we grab the edge and then
+  if (header_builder_.graphid().tileid() != edge_id.tileid() ||
+      header_builder_.graphid().level() != edge_id.level()) {
+    throw std::runtime_error("Tile id or hierarchy level doesn't match");
+  }
+  if (header_builder_.directededgecount() <= edge_id.id()) {
+    throw std::runtime_error(
+        "Given edge doesn't exist: edge id is larger than total edge size in this tile");
+  }
+
+  // get the edges edgeinfo, to do that we need to look at the edgeinfo offset of the edge
+  // so we take the edgeinfo offset from the edge use it as the key into edgeinfo_offset_map_, which
+  // will give us back an edgeinfobuilder
+  const auto& edge = directededges_builder_[edge_id.id()];
+  auto eib = edgeinfo_offset_map_.find(edge.edgeinfo_offset());
+  // bail here if this iterator is the end
+  if (eib == edgeinfo_offset_map_.end()) {
+    throw std::runtime_error("Couldn't find edge info for the given edge: " +
+                             std::to_string(static_cast<int>(edge_id.id())));
+  }
+
+  // we need to use the edgeinfobuilder to add a new tagged value record to it that takes some work
+
+  // first we construct the string that makes up the record we want to store, ours is going to look
+  // like:
+  std::string tagged_value = landmark.to_str();
+
+  // do we already have this record, if so we dont want a copy instead we just want the offset
+    auto name_offset = AddName(tagged_value);
+  
+  // so now we know where we are storing this variable sized string in the tile, we need to keep the
+  // record in the edge info of where that was
+  NameInfo ni{name_offset, 0, 0, 1};
+  eib->second->AddNameInfo(ni);
```

Comment:
after you add landmarks to the edgeinfobuilders, you'll need to update all offsets of structures that come after the edgeinfos in the binary tile, including`header->set_end_offset()`. You can have a look a few lines up in `GraphTileBuilder::AddBins` @vesperlou

Reply:
ok so that wasn't it. of course the StoreTileData function is supposed to update the offset stuff. the problem is here: https://github.com/valhalla/valhalla/blob/75b5e818eb694a47d1b21c5cebf0dad45e476537/src/mjolnir/graphtilebuilder.cc#L362. That `edge_info_offset_` is weird.. it's not an offset, it's a size. IMO the code should be refactored to remove that variable entirely, it doesn't seem to serve any purpose. We can get the size of the edgeinfo's dynamically in StoreTileData. The problem is that var is never updated. Which wasn't a problem so far because no other data builder uses StoreTileData AND does anything to edge infos, so far it's always been the same size. 

Let me draft a PR that'll unblock you here. FWIW, this patch is the main change:
```diff
diff --git a/src/mjolnir/graphtilebuilder.cc b/src/mjolnir/graphtilebuilder.cc
index 4c1fbdb7e..2cd265dac 100644
--- a/src/mjolnir/graphtilebuilder.cc
+++ b/src/mjolnir/graphtilebuilder.cc
@@ -352,14 +352,16 @@ void GraphTileBuilder::StoreTileData() {
     }
 
     // Write the edge data
+    uint32_t edge_info_size = 0;
     header_builder_.set_edgeinfo_offset(header_builder_.complex_restriction_reverse_offset() +
                                         reverse_restriction_size);
     for (const auto& edgeinfo : edgeinfo_list_) {
       in_mem << edgeinfo;
+      edge_info_size += edgeinfo.SizeOf();
     }
 
     // Write the names
-    header_builder_.set_textlist_offset(header_builder_.edgeinfo_offset() + edge_info_offset_);
+    header_builder_.set_textlist_offset(header_builder_.edgeinfo_offset() + edge_info_size);
     for (const auto& text : textlistbuilder_) {
       in_mem << text << '\0';
     }
```

Reply:
other places in the code update that when you add to the edgeinfos. we could so this in addlandmark

Reply:
here you could just call SizeOf before and SizeOf after adding the name info and add the difference to the edge_info_offset_ (which yes is in bytes)

Reply:
here you could just call SizeOf before and SizeOf after adding the name info and add the difference to the `edge_info_offset_` (which yes is in bytes)

Reply:
got it thanks